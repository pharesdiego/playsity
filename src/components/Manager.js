import React, { Component } from 'react';

// this component helps us manage state
class Manager extends Component {
  state = {
    ...this.props.initialState
  }

  updateState = (prop, value) => {
    this.setState({
      [prop]: value
    })
  }

  render() {
    return this.props.render(this.state, this.updateState);
  }
}

export default Manager;