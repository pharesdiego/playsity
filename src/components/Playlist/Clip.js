import React, { Fragment } from 'react';
import styled from 'styled-components';
import Icon from './../Icon';
import { Flex } from './../Styles';
import percentage from './../../utils/percentage';

const Body = styled.div`
  width: 100%;
  height: auto;
  background: ${props => props.isPlaying ? 'linear-gradient(45deg, #882cb7, #8e357a)' : 'rgba(214, 214, 214, 0.22)'};
  padding: 10px 5px;
  position: relative;
  box-shadow: 0px 0px 9px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 10px;
  border-radius: 4px;
  transition: all 0.2s;
`;

const Container = Flex.extend`
  width: 100%;
  height: 100%;
`;

const Section = Flex.extend`
  flex: 1;
`;

const Name = Section.extend`
  text-align: center;
  display: inline;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const ProgressBar = styled.div`
  position: absolute;
  height: 3px;
  bottom: 0%;
  left: 0%;
  background-color: ${props => props.isPlaying ? 'white' : 'rgba(255, 255, 255, 0.40)'};
  transition: 0.2s all;
`;

const Tags = styled.div`
  padding: 5px 20px 0px 5px;
  margin-top: 5px;
  border-top: 1px dashed rgba(255, 255, 255, 0.70);
  display: flex;
`;

const Tag = styled.span`
  max-width: 100px;
  text-align: center;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 4px 5px;
  background-color: rgba(255, 255, 255, 0.15);
  border-radius: 4px;
  &:not(:last-child) {
    margin-right: 5px;
  }
`;
const Clip = (props) => {
  const { name, time, start, end, id, tags: tagsData, isPlaying, onPlay, onDelete, onEdit } = props;
  const tags = tagsData
                .split(/,+/)
                .filter(tag => tag !== '')
                .map((tag, i) => <Tag key={tag+i}>{ tag.trim() }</Tag>);
  return(
    <Body isPlaying={isPlaying}>
      <Container>
        <Name>{ name }</Name>
        <Section>
          {
            id !== null && (
              <Fragment>
                { start + 's' }
                <Icon>chevron_right</Icon>
                { end + 's' }
              </Fragment>
            )
          }
        </Section>
        <Section>
          <Section>
            <Icon onClick={onPlay}>
              play_circle_outline
            </Icon>
          </Section>
          { id !== null && 
            <Fragment>
              <Section>
                <Icon onClick={onEdit}>edit</Icon>
              </Section>
              <Section>
                <Icon onClick={onDelete}>
                  delete_outline
                </Icon>
              </Section>
            </Fragment>
          }
        </Section>
      </Container>
      {
        tags.length > 0 && <Tags>{ tags }</Tags>
      }
      <ProgressBar
        style={{ width: percentage(time, +start, +end) + '%' }}
        isPlaying={isPlaying}
      />
    </Body>
  )
}

export default Clip;