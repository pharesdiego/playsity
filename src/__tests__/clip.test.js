import React from 'react';
import Clip from './../components/Playlist/Clip';
import Manager from './../components/Manager';

const ClipWithManager = (props) => {
  return (
    <Manager
      initialState={{data: {time: '0', isPlaying: false, name: 'test', start: '0', end: '15', tags: 'testing, tag'}}}
      render={(state, updateState) => (
        <Clip
          onPlay={() => updateState('data', {...state.data, isPlaying: true})}
          onEdit={() => updateState('data', {...state.data, name: 'edited name', start: '3', end: '20', tags: 'testing, edited'})}
          {...state.data}
        />
      )}
    />
  )
}

let wrapper;

describe('>>> Clip <<<', () => {
  beforeEach(() => {
    wrapper = mount(<ClipWithManager/>);
  });

  it('Must render', () => {
    expect(wrapper).toBeDefined();
  });

  test('onClick handler must work and update isPlaying state to true', () => {
    wrapper.find('Icon').at(1).simulate('click');
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('onEdit handler must work and edit state', () => {
    wrapper.find('Icon').at(2).simulate('click');
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});