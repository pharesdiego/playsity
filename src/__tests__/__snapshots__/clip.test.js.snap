// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Clip <<< onClick handler must work and update isPlaying state to true 1`] = `
<ClipWithManager>
  <Manager
    initialState={
      Object {
        "data": Object {
          "end": "15",
          "isPlaying": false,
          "name": "test",
          "start": "0",
          "tags": "testing, tag",
          "time": "0",
        },
      }
    }
    render={[Function]}
  >
    <Clip
      end="15"
      isPlaying={true}
      name="test"
      onEdit={[Function]}
      onPlay={[Function]}
      start="0"
      tags="testing, tag"
      time="0"
    >
      <styled.div
        isPlaying={true}
      >
        <div
          className="sc-htpNat Dszuf"
        >
          <styled.div>
            <div
              className="sc-bxivhb bKvnmv"
            >
              <styled.div>
                <div
                  className="sc-EHOje hTYpxf"
                >
                  test
                </div>
              </styled.div>
              <styled.div>
                <div
                  className="sc-ifAKCX cIZoKD"
                >
                  0s
                  <Icon>
                    <styled.div>
                      <div
                        className="sc-bdVaJa qyge"
                      >
                        <i
                          className="material-icons"
                          style={
                            Object {
                              "fontSize": "24px",
                            }
                          }
                        >
                          chevron_right
                        </i>
                      </div>
                    </styled.div>
                  </Icon>
                  15s
                </div>
              </styled.div>
              <styled.div>
                <div
                  className="sc-ifAKCX cIZoKD"
                >
                  <styled.div>
                    <div
                      className="sc-ifAKCX cIZoKD"
                    >
                      <Icon
                        onClick={[Function]}
                      >
                        <styled.div
                          onClick={[Function]}
                        >
                          <div
                            className="sc-bdVaJa qyge"
                            onClick={[Function]}
                          >
                            <i
                              className="material-icons"
                              style={
                                Object {
                                  "fontSize": "24px",
                                }
                              }
                            >
                              play_circle_outline
                            </i>
                          </div>
                        </styled.div>
                      </Icon>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-ifAKCX cIZoKD"
                    >
                      <Icon
                        onClick={[Function]}
                      >
                        <styled.div
                          onClick={[Function]}
                        >
                          <div
                            className="sc-bdVaJa qyge"
                            onClick={[Function]}
                          >
                            <i
                              className="material-icons"
                              style={
                                Object {
                                  "fontSize": "24px",
                                }
                              }
                            >
                              edit
                            </i>
                          </div>
                        </styled.div>
                      </Icon>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-ifAKCX cIZoKD"
                    >
                      <Icon>
                        <styled.div>
                          <div
                            className="sc-bdVaJa qyge"
                          >
                            <i
                              className="material-icons"
                              style={
                                Object {
                                  "fontSize": "24px",
                                }
                              }
                            >
                              delete_outline
                            </i>
                          </div>
                        </styled.div>
                      </Icon>
                    </div>
                  </styled.div>
                </div>
              </styled.div>
            </div>
          </styled.div>
          <styled.div>
            <div
              className="sc-gzVnrw jfohoA"
            >
              <styled.span
                key="testing0"
              >
                <span
                  className="sc-htoDjs ioRxXM"
                >
                  testing
                </span>
              </styled.span>
              <styled.span
                key=" tag1"
              >
                <span
                  className="sc-htoDjs ioRxXM"
                >
                  tag
                </span>
              </styled.span>
            </div>
          </styled.div>
          <styled.div
            isPlaying={true}
            style={
              Object {
                "width": "0%",
              }
            }
          >
            <div
              className="sc-bZQynM gIQXtb"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </styled.div>
        </div>
      </styled.div>
    </Clip>
  </Manager>
</ClipWithManager>
`;

exports[`>>> Clip <<< onEdit handler must work and edit state 1`] = `
<ClipWithManager>
  <Manager
    initialState={
      Object {
        "data": Object {
          "end": "15",
          "isPlaying": false,
          "name": "test",
          "start": "0",
          "tags": "testing, tag",
          "time": "0",
        },
      }
    }
    render={[Function]}
  >
    <Clip
      end="20"
      isPlaying={false}
      name="edited name"
      onEdit={[Function]}
      onPlay={[Function]}
      start="3"
      tags="testing, edited"
      time="0"
    >
      <styled.div
        isPlaying={false}
      >
        <div
          className="sc-htpNat OPVLN"
        >
          <styled.div>
            <div
              className="sc-bxivhb bKvnmv"
            >
              <styled.div>
                <div
                  className="sc-EHOje hTYpxf"
                >
                  edited name
                </div>
              </styled.div>
              <styled.div>
                <div
                  className="sc-ifAKCX cIZoKD"
                >
                  3s
                  <Icon>
                    <styled.div>
                      <div
                        className="sc-bdVaJa qyge"
                      >
                        <i
                          className="material-icons"
                          style={
                            Object {
                              "fontSize": "24px",
                            }
                          }
                        >
                          chevron_right
                        </i>
                      </div>
                    </styled.div>
                  </Icon>
                  20s
                </div>
              </styled.div>
              <styled.div>
                <div
                  className="sc-ifAKCX cIZoKD"
                >
                  <styled.div>
                    <div
                      className="sc-ifAKCX cIZoKD"
                    >
                      <Icon
                        onClick={[Function]}
                      >
                        <styled.div
                          onClick={[Function]}
                        >
                          <div
                            className="sc-bdVaJa qyge"
                            onClick={[Function]}
                          >
                            <i
                              className="material-icons"
                              style={
                                Object {
                                  "fontSize": "24px",
                                }
                              }
                            >
                              play_circle_outline
                            </i>
                          </div>
                        </styled.div>
                      </Icon>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-ifAKCX cIZoKD"
                    >
                      <Icon
                        onClick={[Function]}
                      >
                        <styled.div
                          onClick={[Function]}
                        >
                          <div
                            className="sc-bdVaJa qyge"
                            onClick={[Function]}
                          >
                            <i
                              className="material-icons"
                              style={
                                Object {
                                  "fontSize": "24px",
                                }
                              }
                            >
                              edit
                            </i>
                          </div>
                        </styled.div>
                      </Icon>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-ifAKCX cIZoKD"
                    >
                      <Icon>
                        <styled.div>
                          <div
                            className="sc-bdVaJa qyge"
                          >
                            <i
                              className="material-icons"
                              style={
                                Object {
                                  "fontSize": "24px",
                                }
                              }
                            >
                              delete_outline
                            </i>
                          </div>
                        </styled.div>
                      </Icon>
                    </div>
                  </styled.div>
                </div>
              </styled.div>
            </div>
          </styled.div>
          <styled.div>
            <div
              className="sc-gzVnrw jfohoA"
            >
              <styled.span
                key="testing0"
              >
                <span
                  className="sc-htoDjs ioRxXM"
                >
                  testing
                </span>
              </styled.span>
              <styled.span
                key=" edited1"
              >
                <span
                  className="sc-htoDjs ioRxXM"
                >
                  edited
                </span>
              </styled.span>
            </div>
          </styled.div>
          <styled.div
            isPlaying={false}
            style={
              Object {
                "width": "0%",
              }
            }
          >
            <div
              className="sc-bZQynM kIXdHt"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </styled.div>
        </div>
      </styled.div>
    </Clip>
  </Manager>
</ClipWithManager>
`;
